### ISSUE TITLE
Feature Request: Peering session auto discovery

### ISSUE THREAD

**Issue by FredNeuberger** [respawner/peering-manager#8]
Hi there,

Are there any plans to implement some sort of auto discovery for existing peering sessions?
I wouldn't want to manually create 100, 200 or more sessions.

Cheers

**Comment by respawner** [comment]
Hello,

That's not a feature I thought about first but it could be very useful.

I have no idea, for now, how we could implement it (using NAPALM or SNMP maybe).

**Comment by FredNeuberger** [comment]
@respawner Thanks for the quick reply.
I'm a total newbie to NAPALM, so currently no idea how that would work.
SNMP is probably an option, not sure how easy that is to implement.

What I know is, that parsing config files is working okay. Implemented something 9 years ago in PHP for something similar. Currently no idea how to implement something like this in Python and SQLite, as I have almost zero experience with both of them.

But to keep stuff consistent, it is probably best to check if NAPALM can do the job.

**Comment by FredNeuberger** [comment]
Did some digging and found something that generates output, but my problem is currently the lack of Python knowledge and I can't  do anything with that nested output stuff ...

But the important information is included in the output. Besides the obvious stuff like remote/local-ip/asn, it also includes things like "uptime", "is_up", "description", send/accepted and received prefixes


```
from napalm_base import get_network_driver
driver = get_network_driver('junos')
dev = driver(hostname='host/IP', username='user',password='password')
dev.open()
output=dev.get_bgp_neighbors()
print output
dev.close()
```

http://napalm.readthedocs.io/en/latest/base.html#napalm.base.base.NetworkDriver.get_bgp_neighbors

**Comment by jnikoletich** [comment]
This is what I used to test:

import json
from napalm import get_network_driver
driver = get_network_driver('junos')
dev = driver(hostname='XXX', username='XXX', password='XXX')
dev.open()
bgp_neighbors = dev.get_bgp_neighbors()
dev.close()
print(json.dumps(bgp_neighbors, sort_keys=True, indent=4))

credit = https://networklore.com/napalm-introduction/

**Comment by respawner** [comment]
Nice catch of both of you that was I was investigating on.

So we can use this base of NAPALM code to import known peering sessions. I still have to figure out the UI and the filtering for this.

This code could probably be useful to have a state of the peering sessions known inside Peering Manager.

I'll try to integrate this ASAP.

**Comment by jnikoletich** [comment]
That would be great. I have seen one of our peers have it displayed like t
<img width="420" alt="peering_cap_nobranding" src="https://user-images.githubusercontent.com/29226372/35415788-b7b537b2-01db-11e8-9341-5f825c8af383.PNG">
his

**Comment by FredNeuberger** [comment]
In regards to filtering, it would be probably best to do that via the LAN IP range of the IXP.
Other than that, there is probably just the description or the group the session is configured in (not sure the latter applies to all the vendors).

Again, thank you for your great work so far.

**Comment by jnikoletich** [comment]
@FredNeuberger I think the filtering by IP address would be a good approach. the IX block can be pulled from peeringdb. This will also get rid of the fud as well (transit/ibgp/etc) that is returned via the NAPALM command.

**Comment by respawner** [comment]
The feature has been implemented in master (commit b6e6d34).

It has not been tested though, since I don't have anything to do the testing for now. It needs a router with a supported platform and IX to be imported from PeeringDB.

If anyone is interesting in testing this and reporting if it is working properly that would be quite nice.

**Comment by XioNoX** [comment]
Quick feedback, I'll add more if I have time:
https://github.com/respawner/peering-manager/commit/b6e6d34d5303f24224105f1c50c886537c398887#diff-89106cc0a88a1a94ffbe8ddabfc897fcR418
In get_bgp_sessions,
I had to add `.items()` to `for ip, details in bgp_neighbors[vrf]['peers']:`
(`for ip, details in bgp_neighbors[vrf]['peers'].items():`)
Otherwise it would fail silently.

2nd issue happen when it pulls from the router a peering session whose peer AS# is not in PeeringDB
 - Services internal to the IXP (Equinix route collector uses AS65517)
 - Peers that don't have a PeeringDB page

The error is the following:
```
Request URL: | http://localhost:8000/ix/equinix-chicago/import_from_router
IntegrityError
NOT NULL constraint failed: peering_peeringsession.autonomous_system_id
/home/xionox/.local/lib/python3.6/site-packages/django/db/backends/sqlite3/base.py in execute, line 328
```

`AutonomousSystem.create_from_peeringdb(...)` returns `Null`

Which is used in:
```
values = {
     'autonomous_system': autonomous_system,
     'internet_exchange': self,
    'ip_address': str(bgp_session['ip_address']),
}
peering_session = PeeringSession(**values)
peering_session.save()
```

If I make sure the AS exists:
```
if autonomous_system:
    values = {
        'autonomous_system': autonomous_system,
        'internet_exchange': self,
        'ip_address': str(bgp_session['ip_address']),
    }
    peering_session = PeeringSession(**values)
    peering_session.save()
    number_of_peering_sessions += 1
```
At this point the AS present in PeeringDB and matching peering sessions get created but still show the error:
```
Exception Type: | TypeError
-- | --
success() missing 1 required positional argument: 'message'
/home/xionox/Documents/Projects/github/XioNoX/peering-manager/peering/views.py in process, line 247
```
Which I solved with:
`messages.success(message)` -> `messages.success(request,message)`

Next error then is:
```

Exception Type: | ValueError
-- | --
Cannot assign "<class 'peering.models.AutonomousSystem'>": "UserAction.content_type" must be a "ContentType" instance.
/home/xionox/.local/lib/python3.6/site-packages/django/db/models/fields/related_descriptors.py in __set__, line 216
```
Seems to be related to `UserAction.objects.log_import`
